crates/bevy_graph/src/graphs/map/simple.rs

```diff
@@ -0,0 +1,506 @@
+                        .outgoing_mut()
+                        .remove(&dst);
+                    self.adjacencies
+                        .get_unchecked_mut(dst)
+                        .incoming_mut()
+                        .remove(&src);
+                }
+            }
+            unsafe {
+                self.adjacencies.remove(index).unwrap_unchecked();
+            }
+            self.nodes.remove(index)
+        } else {
+            None
+        }
+    }
+
+    fn remove_edge(&mut self, index: EdgeIdx) -> Option<E> {
+        if let Some(Edge(src, dst, value)) = self.edges.remove(index) {
+            unsafe {
+                self.adjacencies
+                    .get_unchecked_mut(src)
+                    .outgoing_mut()
+                    .remove(&dst);
+                self.adjacencies
+                    .get_unchecked_mut(dst)
+                    .incoming_mut()
+                    .remove(&src);
+            }
+            Some(value)
+        } else {
+            None
+        }
+    }
+
+    fn clear_edges(&mut self) {
+        self.adjacencies
+            .values_mut()
+            .for_each(|map| map.for_each_mut(HashMap::clear));
+        self.edges.clear();
+    }
+
+    fn clear(&mut self) {
+        self.adjacencies.clear();
+        self.edges.clear();
+        self.nodes.clear();
+    }
+
+    #[inline]
+    fn get_node(&self, index: NodeIdx) -> Option<&N> {
+        self.nodes.get(index)
+    }
+
+    #[inline]
+    fn get_node_mut(&mut self, index: NodeIdx) -> Option<&mut N> {
+        self.nodes.get_mut(index)
+    }
+
+    fn find_node(&self, value: &N) -> Option<NodeIdx>
+    where
+        N: PartialEq,
+    {
+        self.nodes
+            .iter()
+            .find_map(|(key, val)| if val == value { Some(key) } else { None })
+    }
+
+    #[inline]
+    unsafe fn get_edge_raw(&mut self, index: EdgeIdx) -> Option<&mut Edge<E>> {
+        self.edges.get_mut(index)
+    }
+
+    #[inline]
+    fn get_edge(&self, index: EdgeIdx) -> Option<EdgeRef<E>> {
+        self.edges.get(index).map(|edge| edge.as_ref_edge())
+    }
+
+    #[inline]
+    fn get_edge_mut(&mut self, index: EdgeIdx) -> Option<EdgeMut<E>> {
+        self.edges.get_mut(index).map(|edge| edge.as_mut_edge())
+    }
+
+    fn find_edge(&self, value: &E) -> Option<EdgeIdx>
+    where
+        E: PartialEq,
+    {
+        self.edges
+            .iter()
+            .find_map(|(key, val)| if &val.2 == value { Some(key) } else { None })
+    }
+
+    fn degree(&self, index: NodeIdx) -> usize {
+        if DIRECTED {
+            self.in_degree(index) + self.out_degree(index)
+        } else {
+            self.adjacencies[index].incoming().len()
+        }
+    }
+
+    type NodeIndices<'n> = slotmap::hop::Keys<'n, NodeIdx, N> where Self: 'n;
+    fn node_indices(&self) -> Self::NodeIndices<'_> {
+        self.nodes.keys()
+    }
+
+    unsafe fn nodes_raw(&self) -> &slotmap::HopSlotMap<NodeIdx, N> {
```

Comment:
Why is this function unsafe? I'm relatively new to rust to rust but from what I understand you only use unsafe for "unsafe superpowers" which I don't think we need here.

Reply:
It's unsafe because you as the user shouldn't play with it. It's needed for the iterator which turns EdgeIdx into EdgeMut because we can't directly have two mut borrows of the graph even if we need different fields. So I did this to expose a field

Reply:
That justification feels more like an argument to have it private, than for it to be unsafe. Unsafe is *strictly* used in Rust for things that can directly cause memory safety problems when used, which doesn't seem to fit here.

Reply:
In my case I'm filtering the graph for some specific nodes. I need the indices later on (to get neighbors of that specific node etc.). I feel like this use case is pretty common (the function hands back an immutable reference so I don't think the user can do anything bad).

Reply:
Yeah i probably should make a sub-trait which is private and exposes the field.
@dmyyy do you think i should add the EdgeIdx into the Edge{Ref/Mut} types?